# <center>《机器学习》— MindSpore 实践练习 </center>

### <center>姓名：XXX  学号：XXXXXXXXXXX</center>

# 4. 基于MindSpore实现梯度下降算法

## 1 实验介绍
        梯度下降是一种非常通用的优化算法，能够为大范围的问题找到最优解，其核心思想是采用不断迭代地方法来调整参数，使得成本函数最小化。梯度下降是优化神经网络时需要计算的，是相对于神经网络权重的成本梯度。很多训练神经网络的算法几乎总是基于使用梯度来使得成本函数下降，因此，梯度下降在神经网络算法原理中起到非常基础和重要的作用。当然还有许多算法对梯度下降思想进行改进和优化，后面也会介绍随机梯度下降算法。同时，梯度下降也可以用来训练线性回归、支持向量机等传统机器学习模型，且当训练集相当大时是常用方法。计算梯度在神经网络中实现会相对复杂一些，但仍然可以很高效且精确的实现。        
    本章实验主要专注于梯度下降的原理，并利用闭式解方程求解梯度。同时，应用MindSpore框架求解梯度。
    本章实验主要分为两个部分：
    （1）使用张量求出解析解。
    （2）使用梯度下降求解。
### 1.1 实验目的
    通过实验了解梯度下降的计算原理，实现求解梯度方法。
### 1.2 实验清单
|  实验  |  简述  |  难度  | 软件环境 | 开发环境 |支持平台|操作系统|
| :-----:| :----: | :----: |:----:   |:----:   |:----:  |:----:  |
| 解析解 | 使用张量实现闭式解求解方法 | 初级 | MindSpore1.8 Python3.7.5 | JupyterNotebook |Ascend/GPU/CPU|Windows/Linux X86|
|方法实现 | 使用梯度下降原理实现方法  | 初级 | MindSpore1.8 Python3.7.5 | JupyterNotebook |Ascend/GPU/CPU|Windows/Linux X86|
### 1.3 开发平台介绍
    MindSpore 最佳匹配昇腾芯片的开源AI计算框架，支持Ascend、GPU、CPU平台。
    MindSpore官网https://www.mindspore.cn ，可下载安装包及查询相关文档。
    Jupyter Notebook是一个基于网页的交互式计算环境，交互式强，易于可视化，适用需要频繁修改的数据分析实验环境。
    ModelArts是面向开发者的一站式AI开发平台，为机器学习与深度学习提供海量数据预处理及半自动化标注、大规模分布式Training、自动化模型生成，及端-边-云模型按需部署能力，帮助用户快速创建和部署模型，管理全周期AI工作流。

## 2 实验环境要求
    在开始本实验前，需要完成实验环境搭建工作。可以搭建Python3.7.5和MindSpore1.8环境完成实验，也可以在华为云ModelArts实验平台完成实验。    
### 2.1 实验环境搭建
    步骤 1	安装miniconda环境。
    从网站上下载miniconda的安装包，我们以Windows环境为主。
    
![](Figures/fig020.png)

![](Figures/fig021.png)

    点击安装miniconda。

![](Figures/fig022.png)

    步骤 2	通过miniconda创建python的程序环境。
    
    安装好的miniconda环境。

![](Figures/fig023.png)

    输入命令创建conda环境，并且使用python3.7.5。

![](Figures/fig024.png)

    创建虚拟环境完成，切换到创建的环境中去，并查看已经安装的Python包。

![](Figures/fig025.png)

![](Figures/fig026.png)

![](Figures/fig027.png)

![](Figures/fig028.png)

    步骤 3  安装Mindspore。

![](Figures/fig029.png)

![](Figures/fig030.png)

    步骤 4  安装jupyter notebook和jupyter lab。

![](Figures/fig031.png)

![](Figures/fig032.png)

![](Figures/fig033.png)

    启动jupyter notebook或者jupyterlab。

### 2.2 创建ModelArts实验环境    
    步骤 1	创建MindSpore开发环境
    进入 ModelArts平台，点击“EI企业智能”，找到AI开发平台ModelArts。
  
![](Figures/fig001.png)

![](Figures/fig002.png)

    进入控制台
    步骤 2	创建ModelArts Notebook
    点击“开发环境”，点击Notebook
    
![](Figures/fig003.png)

    创建新的notebook
![](Figures/fig004.png)

![](Figures/fig005.png)

    可以选择MindSpore1.7.0，python3.7，Ubuntu18.04 的CPU开发环境。
    其他资源选择默认即可。
    
![](Figures/fig006.png)

![](Figures/fig007.png)

    配置好Notebook参数后，点击“立即创建”后，进入Notebook信息预览。确认无误后，点击“立即创建”。
    创建完成后，返回开发环境主界面，页面中显示状态为“创建中”，等待Notebook创建完毕后，当状态变为“运行中”时，可以选择“打开”notebook，进行下一步操作。
    
![](Figures/fig008.png)

![](Figures/fig009.png)

    步骤3 创建新的notebook文件。
    这时可以选择创建MindSpore的notebook。
    点击左上方的文件名"Untitled"，并输入一个与本实验相关的名称，如"softmax"
    
![](Figures/fig010.png)

    步骤 4	在Notebook中编写并执行代码
    在Notebook中，我们输入一个简单的打印语句，然后点击上方的运行按钮，可以查看语句执行的结果：
    
![](Figures/fig011.png)

    步骤 5    保存代码
    代码编写完之后，点击下图所示的“保存”按钮，保存代码和代码执行结果。
    
![](Figures/fig012.png)
    
    步骤 6    从ModelArts下载代码到本地

![](Figures/fig013.png)

    选中需要下载的文件后，在菜单File中选择“Download”，可以将文件下载到本地。
    步骤 7    从本地上传代码到ModelArts
    点击工具栏“上传”按钮，打开上传文件对话框，如下图所示。
    
![](Figures/fig014.png)

![](Figures/fig015.png)

    步骤8 启动和停止
    可以启动notebook，并选择启用时间，到时间后服务会自动停止。
    
![](Figures/fig016.png)

    也可以在启动服务后，选择手动停止服务，点击“停止”按钮。
    
![](Figures/fig017.png)

## 3 实验准备及实验设计
### 3.1 实验介绍
线性回归方程: $\hat y=w_0+w_1𝑥_1+w_2 𝑥_2+…+w_n 𝑥_𝑛$

通常会采用MSE均方根误差作为成本函数：$h_w=1/𝑚 \sum_{𝑖=1}^𝑚(w^𝑇 𝑥^{(𝑖)}−𝑦^{(𝑖)})^2 $

* (1) 求解析解，线性回归的解可以用一个公式简单地表达出来， 这类解叫作解析解(analytical solution)，可以直接计算得到$\hat w =(𝑋^𝑇 𝑋)^{−1} 𝑋^𝑇 𝑦 $
* (2) 采用梯度下降方法，通过梯度向量寻优，梯度向量乘以 $\eta$ 用于确定下坡步长的大小 $w^{(next step)}= w -\eta\nabla_w MSE(w)$

参考mindspore 仓库：https://github.com/LZU-Machine-learning-theory/MindSpore_case/tree/main/case4

### 3.2 实验目标
* 学会闭式解求解推导过程。
* 掌握闭式解求解实现流程。
* 学会梯度求解的推导过程。
* 掌握梯度下降的实现流程。
### 3.3 实验总体设计
1. 导入实验所需模块
2. 导入使用数据
3. 实现方法
4. 查看结果    
    导入实验所需模块：该步骤通常都是程序编辑的第一步，将实验代码所需要用到的模块包用import命令进行导入。
    导入数据并预处理：这里给出已有数据，并不需要做额外处理。
    实现方法：利用mindspore中的Tensor张量进行闭式解求解。
    查看结果：传入数据，调用方法计算结果并查看。
    
### 3.4 实验过程
#### 3.4.1 导入所需模块
    这里需要导入相关依赖库:
    import numpy as np
    import matplotlib.pyplot as plt
    from mindspore import Tensor
    Numpy模块主要用于数据的基本运算操作。Matplotlib模块主要用于画图。MindSpore中的Tensor用于实现梯度下降的张量。  
#### 3.4.2 导入实验数据
    这里给出测试数据点，如下所示：
    x1 = np.array([55,71,68,87,101,87,75,78,93,73])    # 给出x的值
    y1 = np.array([91,101,87,109,129,98,95,101,104,93])# 给出y的值

## 4. 实验过程及内容
### 4.1 使用张量求出解析解


1. 首先需要确定成本函数，这里使用MSE作为成本函数
$$MSE = (X, h_0)=\frac{1}{m}\sum ^m_{i=1}(W^Tx^{(i)}-y^{(i)})^2$$
也可以简写成：
$$MSE =\frac{1}{m}\sum (y-WX)^2$$

2. 求解目标——得到成本函数的最小值
$$target = min ||y-WX||^2 = min (y-WX)^T(y-WX)$$

3. 将求解公式展开得到：
$$target = min (y-WX)^T(y-WX)= W^TX^TWX-y^TWX-yW^TX^T+y^Ty$$

4. 对W求偏导
$$\frac{\partial target}{\partial W}= \frac{\partial(W^TX^TWX-y^TWX-yW^TX^T+y^Ty)}{\partial W}=2X^TXW-2X^Ty+0=0$$

根据$$\frac{\partial W^TX}{\partial W}=\frac{\partial X^TW}{\partial W}=X$$
$$\frac{\partial W^TXW}{\partial W}=XW + X^TW$$
矩阵对称则有：$$XW + X^TW=2XW$$

5. 求解W
$$2X^TXW-2X^Ty=0$$
$$2X^TXW=2X^Ty$$
$$W=(X^TX)^{-1}X^Ty$$



#### 4.1.1 使用numpy测试闭式解求解方法
    测试代码如下：

In [99]:
# 采用解析解的方式计算可得到：
import numpy as np
    
x1 = np.array([55,71,68,87,101,87,75,78,93,73])    # 给出x的值
y1 = np.array([91,101,87,109,129,98,95,101,104,93])# 给出y的值
x_b = np.c_[np.ones((10, 1)), x1]                  # 在x中增加一列，值均为1

w_best_np = np.linalg.inv(x_b.T.dot(x_b)).dot(x_b.T).dot(y1)  # 按公式计算得到w，np.linalg.inv()是对矩阵求逆
print("w_best_np:",w_best_np)                     #  打印输出结果w


w_best_np: [44.25604341  0.7175629 ]


#### 4.1.2 采用张量求出解析解
    数据准备，使用numpy生成数据，之后使用mindspore的tensor进行转换：这里使用数据和前面的相同。

In [90]:
import numpy as np
import matplotlib.pyplot as plt
from mindspore import Tensor
def data_generate():                                  # 定义数据生成函数
    x = np.array([55,71,68,87,101,87,75,78,93,73])    # 给出x的值
    y = np.array([91,101,87,109,129,98,95,101,104,93])# 给出y的值

    x = Tensor(x.astype(np.float32))                  # 将x转换为Tensor
    y = Tensor(y.astype(np.float32))                  # 将y转换为Tensor
   
    return x,y

In [93]:
def ols_algebra(x, y):
    '''
    解析解
    '''
    #根据解析计算方法求解w，补充代码
    

    return w1,w0 
    pass

    通过填充后的代码查看结果。

In [94]:
#按要求补充代码

    #通过填充后的代码查看结果
    x,y = data_generate()
    w1,w0 = ols_algebra(x,y)
    print(w0)
    print(w1)
    #参考输出：
    44.256042
    0.7175629

### 4.2 采用梯度下降求解
    实现梯度下降，需要计算模型关于参数$w_j$的成本函数的梯度。也就是说，需要计算改变$w_j$，成本函数会改变多少。这里采用偏导数计算:
$$\frac{\partial}{\partial w_j}MSE(W)=\frac{2}{m}\sum ^m_{i=1}(W^Tx^{(i)}-y^{(i)})x^{(i)}_j$$
    其中，$m$是样本个数。
    计算成本函数的梯度向量：
    $$\nabla _W MSE(W)=\begin{pmatrix} \frac{\partial}{\partial W_0}MSE(W)\\ \frac{\partial}{\partial W_1 }MSE(W)\\ ...\\ \frac{\partial}{\partial W_n}MSE(W) \end{pmatrix}=\frac{2}{m}X^T(xW-y)$$
    
    梯度下降的步骤为：
$$W^{(next)}=W-\eta \nabla _W MSE(W)$$

#### 4.2.1 采用张量实现梯度下降方法
    根据梯度下降的公式，更新梯度，在给定迭代次数(500-15000)，学习率0.00001的条件下,验证方法的使用效果
    需要补充下列代码：

In [95]:
def ols_gradient_descent(x,y,lr,num_iter):
    '''
    梯度下降解
    lr: 学习率
    num_iter: 迭代次数
   '''
    return w1,w0
    pass
    

In [4]:
#按要求补充代码

#### 4.2.2  结合补充后的代码进行绘图结果比较

In [5]:
def plot_pic(w1,w0,w1_,w0_,x,y):
    '''
    画图
    '''
    fig, axes = plt.subplots(1,2, figsize=(15,5))  # 采用subplots绘制子图1*2个15x5大小的子图
    w1 = w1.asnumpy()                              # 返回具有从该数组复制的值的 numpy.ndarray 对象
    w0 = w0.asnumpy()
    w1_ = w1_.asnumpy()
    w0_ = w0_.asnumpy()
    x = x.asnumpy()
    y = y.asnumpy()

    axes[0].scatter(x,y)                           # 绘制y与x的散点图，并使用不同的标记大小或颜色
    axes[0].plot(np.array([50,110]), np.array([50,110])*w1 + w0, 'r') #根据给出的x和y画出红线
    axes[0].set_title("OLS")

    axes[1].scatter(x,y)
    axes[1].plot(np.array([50,110]), np.array([50,110])*w1_ + w0_, 'r')
    axes[1].set_title("Gradient descent")

    plt.show()

#### 4.2.3  调用梯度下降方法执行结果

    x,y = data_generate()          # 调用data_generate函数生成数据

    w1,w0 = ols_algebra(x,y)       # 调用ols_algebra函数计算解析解
    print(w1)
    print(w0)

    w1_,w0_ = ols_gradient_descent(x,y,lr = 0.00001, num_iter = 500) # 调用ols_gradient_descent函数使用梯度下降方法求解w
    print(w1_)
    print(w0_)

    plot_pic(w1,w0,w1_,w0_,x,y)

    w1_,w0_ = ols_gradient_descent(x,y,lr = 0.00001, num_iter = 15000) # 增加迭代次数后查看求解w
    print(w1_)
    print(w0_)

    plot_pic(w1,w0,w1_,w0_,x,y)
    # 参考输出：
    0.7175629
    44.256042
    1.2633123
    0.12807482
![](Figures/fig018.png)
    
    1.2244889
    3.2672513
![](Figures/fig019.png)

## 5 实验总结
    本实验提供了一个基于MindSpore的张量实现梯度下降闭式解实验,展示了如何通过张量构建闭式解的流程。实验主要对梯度下降进行实现，在学习了直接求解过程后，利用MindSpore提供的张量Tensor完成梯度下降方法的实现并使用提供数据进行验证绘图。